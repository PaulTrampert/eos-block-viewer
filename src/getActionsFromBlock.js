export default (block) => {
  let actions = [];
  for(let transaction of block.transactions) {
    if (transaction.trx.transaction) {
      actions = [...actions, ...transaction.trx.transaction.actions];
    }
  }
  return actions;
};