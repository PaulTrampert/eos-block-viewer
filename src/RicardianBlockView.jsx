import React from 'react';
import PropTypes from 'prop-types';
import Spinner from 'react-bootstrap/Spinner';
import api from './apiWrapper';
import getActionsFromBlock from './getActionsFromBlock';
import RicardianTransaction from './RicardianTransaction';
import styles from './RicardianBlockView.module.css';


class RicardianBlockView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      abis: null
    }
  }

  componentDidMount = async () => {
    const {block} = this.props;
    const actions = getActionsFromBlock(block);
    const abis = {};
    for(let action of actions) {
      if (!abis[action.account]) {
        abis[action.account] = await api.getAbi(action.account);
      }
    }
    this.setState({abis});
  }

  render = () => {
    const {block} = this.props;
    const {abis} = this.state;
    const transactions = block.transactions
      .filter(t => t.trx.id)
      .map(t => t.trx.transaction)
    if (abis) {
      return (
        <>
          {
            abis &&
            transactions.map((t, idx) => <RicardianTransaction key={idx} transaction={t} abis={abis} />)
          }
        </>
      );
    }
    return (
      <div className={styles.loading}>
        <Spinner animation="border" role="status">
          <span className="sr-only">Loading...</span>
        </Spinner>
      </div>
    );
  }
}

RicardianBlockView.propTypes = {
  block: PropTypes.object
};

export default RicardianBlockView;