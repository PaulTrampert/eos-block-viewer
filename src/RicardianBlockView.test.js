import React from 'react';
import {shallow} from 'enzyme';
import api from './apiWrapper';
import getActionsFromBlock from './getActionsFromBlock';
import RicardianBlockView from './RicardianBlockView';

jest.mock('./apiWrapper');
jest.mock('./getActionsFromBlock');

describe('RicardianBlockView', () => {
  let subject;
  let instance;
  let block;
  let actions;

  beforeEach(() => {
    actions = [
      {account: 'a'},
      {account: 'b'},
      {account: 'b'}
    ];

    block = {
      transactions: [
        {
          trx: {
            id: 'asdf',
            transaction: {}
          }
        },
        {
          trx: {
            id: ';lkj',
            transaction: {}
          }
        }
      ]
    };

    api.getAbi.mockImplementation((account) => ({account, someprop: 'bloop'}));
    getActionsFromBlock.mockReturnValue(actions);

    subject = shallow(<RicardianBlockView block={block} />);
    instance = subject.instance();
  });

  describe('componentDidMount', () => {
    beforeEach(() => {
      api.getAbi.mockClear();
    });

    it('gets the actions from the block', async () => {
      await instance.componentDidMount();

      expect(getActionsFromBlock).toHaveBeenCalledWith(block);
    });

    it("gets the abi once for each unique action account", async () => {
      await instance.componentDidMount();

      expect(api.getAbi.mock.calls).toEqual([["a"], ["b"]]);
    });

    it('saves the abis to state', async () => {
      await instance.componentDidMount();

      expect(subject.state().abis).toEqual({
        'a': {account: 'a', someprop: 'bloop'},
        'b': {account: 'b', someprop: 'bloop'}
      });
    });
  });

  describe('render', () => {
    describe("when abis have not been loaded", () => {
      it('renders the loading spinner', () => {
        expect(subject).toMatchSnapshot();
      });
    });

    describe('when abis have been loaded', () => {
      beforeEach(() => {
        subject.setState({
          abis: {
            'a': {},
            'b': {}
          }
        });
      });

      it('renders each transaction in a RicardianTransaction view', () => {
        expect(subject).toMatchSnapshot();
      });
    });
  });
});