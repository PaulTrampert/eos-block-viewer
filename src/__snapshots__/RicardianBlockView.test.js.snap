// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RicardianBlockView render when abis have been loaded renders each transaction in a RicardianTransaction view 1`] = `
<Fragment>
  <RicardianTransaction
    abis={
      Object {
        "a": Object {},
        "b": Object {},
      }
    }
    key="0"
    transaction={Object {}}
  />
  <RicardianTransaction
    abis={
      Object {
        "a": Object {},
        "b": Object {},
      }
    }
    key="1"
    transaction={Object {}}
  />
</Fragment>
`;

exports[`RicardianBlockView render when abis have not been loaded renders the loading spinner 1`] = `
<div
  className="loading"
>
  <Spinner
    animation="border"
    role="status"
  >
    <span
      className="sr-only"
    >
      Loading...
    </span>
  </Spinner>
</div>
`;
